File: java/src/main/java/io/github/flashlock/KahnQueue/NodeMachine.java
package io.github.flashlock.KahnQueue;

import io.github.flashlock.utils.StateMachine;

import java.util.Map;
import java.util.Set;

final class NodeMachine extends StateMachine<NodeState> {
  private static final Map<NodeState, Set<NodeState>> NODE_TRANSITIONS = Map.of(
      NodeState.QUEUED, Set.of(NodeState.READY, NodeState.PRUNED),
      NodeState.READY, Set.of(NodeState.ACTIVE, NodeState.PRUNED),
      NodeState.ACTIVE, Set.of(NodeState.COMPLETE, NodeState.PRUNED),
      NodeState.COMPLETE, Set.of(),
      NodeState.PRUNED, Set.of()
  );

  int numSources;
  int id;

  private NodeMachine(NodeState initialState, Map<NodeState, Set<NodeState>> transitions) {
    super(initialState, transitions);
  }

  public static NodeMachine create(int id, int numSources) {
    var machine = new NodeMachine(NodeState.QUEUED, NODE_TRANSITIONS);
    machine.numSources = numSources;
    machine.id = id;
    machine.tryReady();
    return machine;
  }

  @Override
  public boolean canTransition(NodeState to) {
    if (is(NodeState.QUEUED) && numSources > 0 && to == NodeState.READY) return false;
    return super.canTransition(to);
  }

  public void decrement() {
    if (numSources == 0) {
      throw new IllegalStateException("Attempting to decrement below zero");
    }
    numSources--;
    tryReady();
  }

  private void tryReady() {
    if (canTransition(NodeState.READY)) {
      transition(NodeState.READY);
    }
  }
}

File: go/kahnqueue/node_machine.go
package kahnqueue

import (
	"errors"
	"fmt"

	"github.com/Flashlock/kahn-queue/go/statemachine"
)

// NodeMachine embeds a [statemachine.Machine] for node lifecycle plus Kahn predecessor counts.
type NodeMachine struct {
	*statemachine.Machine[NodeState]
	ID         int
	NumSources int
}

// NewNodeMachine creates a machine for id with numSources incoming edges.
func NewNodeMachine(id, numSources int) *NodeMachine {
	m := &NodeMachine{
		Machine:    statemachine.New(NodeQueued, nodeTransitionTable()),
		ID:         id,
		NumSources: numSources,
	}
	m.tryReady()
	return m
}

// CanTransition applies the Java NodeMachine guard, then the transition table.
func (m *NodeMachine) CanTransition(to NodeState) bool {
	if m.Is(NodeQueued) && m.NumSources > 0 && to == NodeReady {
		return false
	}
	return m.Machine.CanTransition(to)
}

// Transition moves to to if allowed (including NodeMachine-specific rules).
func (m *NodeMachine) Transition(to NodeState) error {
	if !m.CanTransition(to) {
		return fmt.Errorf("invalid transition: %v → %v", m.State(), to)
	}
	m.Machine.SetStateDirect(to)
	return nil
}

// Decrement reduces remaining unsatisfied predecessors; may transition to READY.
func (m *NodeMachine) Decrement() error {
	if m.NumSources == 0 {
		return errors.New("Attempting to decrement below zero")
	}
	m.NumSources--
	m.tryReady()
	return nil
}

func (m *NodeMachine) tryReady() {
	if m.CanTransition(NodeReady) {
		_ = m.Transition(NodeReady)
	}
}

func (m *NodeMachine) forceState(s NodeState) {
	m.Machine.SetStateDirect(s)
}
